python-package/xgboost/dask.py

```diff
@@ -1,121 +1,589 @@
+        else:
+            labels = None
+        if self.has_weights:
+            weights = concat(weights)
+        else:
+            weights = None
+
+        dmatrix = DMatrix(data,
+                          labels,
+                          weight=weights,
+                          missing=self._missing,
+                          feature_names=self._feature_names,
+                          feature_types=self._feature_types)
+        return dmatrix
+
+    def get_worker_data_shape(self, worker):
+        '''Get the shape of data X in each worker.'''
+        data, _, _ = self.get_worker_parts(worker)
+
+        shapes = [d.shape for d in data]
+        rows = 0
+        cols = 0
+        for shape in shapes:
+            rows += shape[0]
+            cols += shape[1]
+        return (rows, cols)
+
+    def num_row(self):
+        return self.n_rows
+
+    def num_col(self):
+        return self.n_cols
+
+
+def _get_rabit_args(worker_map, client):
+    '''Get rabit context arguments from data distribution in DaskDMatrix.'''
+    host = distributed_comm.get_address_host(client.scheduler.address)
+
+    env = client.run_on_scheduler(_start_tracker, host.strip('/:'),
+                                  len(worker_map))
+    rabit_args = [('%s=%s' % item).encode() for item in env.items()]
+    return rabit_args
+
+# train and predict methods are supposed to be "functional", which meets the
+# dask paradigm.  But as a side effect, the `evals_result` in single-node API
+# is no longer supported since it mutates the input parameter, and it's not
+# intuitive to sync the mutation result.  Therefore, a dictionary containing
+# evaluation history is instead returned.
+
+
+def train(client, params, dtrain, *args, evals=(), **kwargs):
+    '''Train XGBoost model.
+
+    parameters:
+    ----------
+    client: dask.distributed.Client
+      Specify the dask client used for training.  Use default client
+      if it's None.
+
+    Other parameters are the same as `xgboost.train` except for `evals_result`
+    is returned as part of function return value instead of argument.
+
+    Returns
+    -------
+    results: dict
+      A dictionary containing trained booster and evaluation history.
+     `history` field is the same as `eval_result` from `xgboost.train`.
+
+      .. code-block:: python
+
+        {'booster': xgboost.Booster,
+         'history': {'train': {'logloss': ['0.48253', '0.35953']},
+                     'eval': {'logloss': ['0.480385', '0.357756']}}}
+
+    '''
+    _assert_dask_installed()
+    if platform.system() == 'Windows':
+        msg = 'Windows is not officially supported for dask/xgboost,'
+        msg += ' contribution are welcomed.'
+        logging.warning(msg)
+
+    if 'evals_result' in kwargs.keys():
+        raise ValueError(
+            'evals_result is not supported in dask interface.',
+            'The evaluation history is returned as result of training.')
+
+    if client is None:
```

Comment:
The code in `predict` uses `if not client`. Can we make these identical?

Also, if I call `xgb.dask.train(params, dtrain)` [*] will this trigger the `get_client()` call? Or do I need to explicitly write `xgb.dask.train(None, params, dtrain)`?

If would be good to support (*) in addition to `xgb.dask.train(client, params, dtrain)`. Should we just make this a keyword argument?

Reply:
This amounts to an API that behaves like this:
```
xgb.dask.train(params, dtrain)
```

or

```
xgb.dask.train(params, dtrain, client=my_client)
```

If we don't like the look of the position in the latter, I'm fine with leaving it as it is.

Reply:
@mt-jones I tried this.  There are other arguments defined as `*args` and passed into `xgboost.train`.  I don't feel comfortable adding an extra at the middle of `*args` and `**kwargs`.  It's quite confusing to me along with duck typing ...

Reply:
But will address other comments.